src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs

```diff
@@ -421,6 +421,36 @@ internal abstract class PEModuleBuilder<TCompilation, TSourceModuleSymbol, TAsse
 
         public abstract TEmbeddedTypesManager EmbeddedTypesManagerOpt { get; }
 
+        private bool _needsNonNullTypesAttribute;
+        /// <summary>
+        /// Track usage of the NonNullTypesAttribute injected type, so it can be added to module to be emitted.
+        /// </summary>
+        protected bool NeedsNonNullTypesAttribute
+        {
+            get => _needsNonNullTypesAttribute;
+            set
+            {
+                Debug.Assert(!InjectedSymbolsAreFrozen);
+                _needsNonNullTypesAttribute = value;
```

Comment:
>_needsNonNullTypesAttribute = value; [](start = 16, length = 36)

I am not sure if a writable property is the right API shape, I assume we are not expecting anyone arbitrarily changing the value. One should only be able to record that the type is needed and never erase this fact.